pkg/virt-controller/watch/vm/vm.go

```diff
@@ -3289,6 +3280,69 @@ func (c *Controller) sync(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineI
+			log.Log.Object(vm).Infof("failed to store Instancetype ControllerRevisions for VirtualMachine: %s/%s", vm.Namespace, vm.Name)
+			c.recorder.Eventf(vm, k8score.EventTypeWarning, common.FailedCreateVirtualMachineReason, "Error encountered while storing Instancetype ControllerRevisions: %v", err)
+			return vm, common.NewSyncError(fmt.Errorf("Error encountered while storing Instancetype ControllerRevisions: %v", err), common.FailedCreateVirtualMachineReason)
+		}
+	case virtv1.Expand, virtv1.ExpandAll:
+		if !shouldExpandInstancetypeAndPreference(vm, referencePolicy) {
+			break
+		}
+
+		expandVMCopy, err := c.instancetypeMethods.Expand(vm, c.clusterConfig)
+		if err != nil {
+			return vm, common.NewSyncError(fmt.Errorf("error encountered while expanding instance type into VirtualMachine: %v", err), common.FailedCreateVirtualMachineReason)
+		}
+
+		// Only update the VM if we have changed something by applying an instance type and preference
+		if !equality.Semantic.DeepEqual(vm.Spec, expandVMCopy.Spec) {
+			updatedVm, err := c.clientset.VirtualMachine(expandVMCopy.Namespace).Update(context.Background(), expandVMCopy, metav1.UpdateOptions{})
+			if err != nil {
+				return vm, common.NewSyncError(fmt.Errorf("error encountered when trying to update VirtualMachine with expanded instance type and preference: %v", err), FailedUpdateErrorReason)
+			}
+			// Return at this point as the update will trigger another sync
+			return updatedVm, nil
+		}
+	}
+
+	// If we have ControllerRevisions make sure they are fully up to date before proceeding
+	if err := c.instancetypeMethods.Upgrade(vm); err != nil {
+		log.Log.Object(vm).Reason(err).Errorf("failed to upgrade instancetype.kubevirt.io ControllerRevisions for VirtualMachine: %s/%s", vm.Namespace, vm.Name)
+		c.recorder.Eventf(vm, k8score.EventTypeWarning, common.FailedCreateVirtualMachineReason, "error encountered while upgrading instancetype.kubevirt.io ControllerRevisions: %v", err)
+		return vm, common.NewSyncError(fmt.Errorf("error encountered while upgrading instancetype.kubevirt.io ControllerRevisions: %v", err), common.FailedCreateVirtualMachineReason)
+	}
+	return vm, nil
+}
+
+func shouldExpandInstancetypeAndPreference(vm *virtv1.VirtualMachine, referencePolicy virtv1.InstancetypeReferencePolicy) bool {
+	// With Expand we only want to expand if we are using instance types and preferences with no revisionNames set
+	if referencePolicy == virtv1.Expand {
+		if vm.Spec.Instancetype == nil && vm.Spec.Preference == nil {
+			return false
+		}
+		if vm.Spec.Instancetype != nil && vm.Spec.Instancetype.RevisionName != "" {
+			log.Log.Object(vm).Infof("not expanding as instance type already has revisionName")
+			return false
+		}
+		if vm.Spec.Preference != nil && vm.Spec.Preference.RevisionName != "" {
+			log.Log.Object(vm).Infof("not expanding as preference already has revisionName")
+			return false
+		}
+	}
+	// Otherwise for ExpandAll we always expand
```

Comment:
What happens if you're using the default policy, then:
- A VM gets created and the `revisionName` gets populated
- You switch the policy to `ExpandAll` and things get expanded in the VM
- You switch back to the default policy

I'm worried that, at that point, the VM would be in a weird state where it's tracking an InstanceType but also has expanded values. Hopefully that's not a valid concern, but if it is maybe we need to remove the revision name when we expand.

Either way, `ExpandAll` doesn't seem that popular/necessary, maybe we start without it and add it later if needed?

Reply:
`expand`and `expandAll` both clear `spec.{instancetype,preference}` entirely so in your scenario there is no reference through either a name or CR after expansion and so no impact when the config returns to `reference`.

Reply:
Oh perfect, well then
/lgtm